packages/block-library/src/image/edit.native.js

```diff
@@ -100,6 +103,12 @@ class ImageEdit extends React.Component {
 		this.removeMediaUploadListener();
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		this.setState( ( state ) => ( {
+			isCaptionSelected: nextProps.isSelected && state.isCaptionSelected,
+		} ) );
```

Comment:
I am guarding the `isCaptionSelected` state here with the image block's `isSelected` prop in order to avoid a noticeable flicker in the toolbar buttons when changing to another block with buttons. There seems to be a lag between the image block's `isSelected` prop being updated to false and its `isCaptionSelected` state being updated to false by the `onCaptionBlur` function (which is passed as the caption's `onBlur` prop). That lag created a split second where the toolbar buttons for both the caption and the newly-selected block were shown at the same time.

My original approach was to just set the caption's `isSelected` prop in the image block's render function as `this.props.isSelected && this.state.isCaptionSelected`. But this did not work when:

(1) the caption was selected (so the image's `isCaptionSelected` state was true);
(2) the caption had no text; and
(3) another block was selected, causing the image's `isSelected` prop to change to false.

When the new block is selected in (3), the caption's `onBlur` prop function (the `onCaptionBlur` function) would _not_ be called and update the image block's `isCaptionSelected` state to be false becasuse an _empty_ caption is not even rendered once the image block's `isSelected` prop is false:
https://github.com/WordPress/gutenberg/blob/a738017ed8dfbd968c4fc759cd67c63db38cffcd/packages/block-library/src/image/edit.native.js#L420-L422

Because the `isCaptionSelected` state is never updated to false, if the user re-selects the image with the empty caption, the empty caption would appear _and immediately be selected_ even though the user did not tap on the caption (the user couldn't have directly tapped on the caption because empty captions are not shown until the related image is selected). If the caption was not empty, then this bug does not appear because the caption would always render, allowing its `onBlur` prop function to update the `isCaptionSelected` state to false.

Reply:
Thanks for the explanation!
Maybe for this one it would be worth to add a short comment in code?
So someone else (or future us) that sees it will understand the intention easily.

Reply:
Good idea. I included this info in the commit body, but it makes sense to add a short explanation in a comment here. 

I pushed an update addressing this.